libgadget/sfr_eff.c

```diff
@@ -476,34 +454,7 @@ static void cooling_relaxed(int i, double egyeff, double dtime, double trelax) {
     const double densityfac = pow(SPH_EOMDensity(i) * All.cf.a3inv, GAMMA_MINUS1) / GAMMA_MINUS1;
     double egycurrent = SPHP(i).Entropy *  densityfac;
 
-    if(All.BlackHoleOn && SPHP(i).Injected_BH_Energy > 0)
-    {
-        double redshift = 1./All.Time - 1;
-        struct UVBG uvbg = get_local_UVBG(redshift, P[i].Pos);
-        egycurrent += SPHP(i).Injected_BH_Energy / P[i].Mass;
-
-        const double u_to_temp_fac = (4 / (8 - 5 * (1 - HYDROGEN_MASSFRAC))) * PROTONMASS / BOLTZMANN * GAMMA_MINUS1
-        * All.UnitEnergy_in_cgs / All.UnitMass_in_g;
-
-        double temp = u_to_temp_fac * egycurrent;
-
-        if(temp > 5.0e9)
-            egycurrent = 5.0e9 / u_to_temp_fac;
-
-        if(egycurrent > egyeff)
-        {
-            double ne = SPHP(i).Ne;
-            double tcool = GetCoolingTime(redshift, egycurrent, SPHP(i).Density * All.cf.a3inv, &uvbg, &ne, SPHP(i).Metallicity);
```

Comment:
trelax is modified here. I don't think you still capture this logic after the removal.. You would compute trelax the regular way from two phase model. Here trelax is set to tcool if the BH makes the particles too hot.  I am not sure about the importance of doing this.

But I also don't quite understand the purpose of having this trelax -- I don't think I found this treatment in the SH03 multi-phase paper. Did you understand this part of the model?

Reply:
No, I don't understand this part of the model! It is a bit mysterious. 

Hotter gas generally has a longer cooling time, but trelax is only modified if the new cooling time is *shorter* than the old one. I don't think that would ever happen if the BH heats the gas. So the new trelax I think doesn't do much?

It looks like the code is trying very hard to prevent the black hole feedback from shutting down star formation in star forming regions, but I don't understand why.

Reply:
- tcool is compared against treleax, which seems to be mostly come from tsfr. So this is sort of deciding if sfr or BH is dominating the entropy change.

- egycurrent can be so much less than egyeff that after injecting BH feedback it is still less.

- also looks like the injected energy is not conserved, since we only use it in the first step, and then clears it.

Notice that the _eff part of the file name says this is about effective eq of state, so it makes sense the eq of state depends on if the particle is in the BH feedback region, which introduces an explicit coupling with the BH module. 

Is there a better way to model this coupling than passing in this weird injected_BH_energy? Is it possible to combine it with other transient variables you introduced, or use a similar technique to lift it out of SphP?

Reply:
tsfr notice is not coming from egyeff, but from the free-fall time based on the density. So the code is finding places where the effective cooling time is shorter than the free-fall time...but the effective equation of state is supposed to avoid that.

I *think* that this part of the model is trying to allow for the black hole feedback to stimulate star formation in already star forming regions. But I could be wrong.

It is absolutely possible to move injected_BH_energy to sph_scratch_data. I actually wrote that code first. I'm just not sure the coupling makes that much sense: upstream doesn't seem to do it.